### PR TITLE
A torchtext tutorial to pre-process a non-built-in dataset

### PR DESCRIPTION
**Pull request by anp-scp**
This tutorial illustrates the usage of torchtext (0.15.0) on a dataset that is not a built-in dataset in torchtext.

This tutorial shows how to:

1. Read a dataset using Torchdata 0.6.0
2. Tokenize sentence using torchtext
3. Apply transforms to sentence using torchtext
4. Perform bucket batching using torchtext

cc @pytorch/team-text-core @Nayef211

### CODE REVIEW

**Comment by anp-scp** [comment]
Hi @pytorch/team-text-core,  @Nayef211,

I request you to kindly review and approve the pull request. If you have any feedback or suggestions, I would be grateful to hear them.

Best Regards

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+In this tutorial, we will learn how to:
+
+* Read a dataset
+* Tokenize sentence
+* Apply transforms to sentence
+* Perform bucket batching
+
+Let us assume that we need to prepare a dataset to train a model that can perform English to
+German translation. We will use a tab-delimited German - English sentence pairs provided by
+the `Tatoeba Project <https://tatoeba.org/en>`_ which can be downloaded from this link: `Click
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:24]: nit: can we just add the download link to the "this link" text? "Click Here" comes off a bit phishy.

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+Here <https://www.manythings.org/anki/deu-eng.zip>`__.
+
+Sentence pairs for other languages can be found in this link:
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:27]: nit: same comment as above. Let's hyperlink the "this link" text and get rid of the line below.

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+#
+# First, download the dataset, extract the zip, and note the path to the file `deu.txt`.
+#
+# Ensure that following packages are installed:
+#
+# * `Torchdata 0.6.0 <https://pytorch.org/data/beta/index.html>`_ (Installation instructions: `C\
+#   lick here <https://github.com/pytorch/data>`__)
+# * `Torchtext 0.15.0 <https://pytorch.org/text/stable/index.html>`_ (Installation instructions:\
+#   `Click here <https://github.com/pytorch/text>`__)
+# * Spacy (Docs: `Click here <https://spacy.io/usage>`__)
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:44]: nit: same as above. Get rid of "Click here" and hyperlink the text directly

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+eng = spacy.load("en_core_web_sm") # Load the English model to be used for tokenizing
+de = spacy.load("de_core_news_sm") # Load the German model to be used for tokenizing
+
+# %%
+# Now we will load the dataset
+
+FILE_PATH = 'data/deu.txt'
+dataPipe = dp.iter.IterableWrapper([FILE_PATH])
+dataPipe = dp.iter.FileOpener(dataPipe, mode='rb')
+dataPipe = dataPipe.parse_csv(skip_lines=0, delimiter='\t', as_tuple=True)
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:72]: nit: let's use snake case for variable names to be consistent with other tutorials. Change `datePipe` to `data_pipe`

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+# 2. At line 3, we pass the iterable to `FileOpener` which then
+#    opens the file in read mode
+# 3. At line 4, we call a function to parse the file, which
+#    again returns an iterable of tuples representing each rows
+#    of the tab-delimited file
+#
+# Data pipes can be thought of something like a dataset object, on which
+# we can perform various operations.
+# Check `this tutorial <https://pytorch.org/data/beta/dp_tutorial.html>`_ for more details on
+# data pipes.
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:87]: Change to DataPipes

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+for sample in dataPipe:
+    print(sample)
+    break
+
+# %%
+# Note that we also have attribution details along with pair of sentences. We will
+# write a small function to remove the attribution details:
+
+def remove_attribution(row):
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:100]: nit: let's use camel case for function names to be consistent with other tutorials. Change `remove_attribution` to `removeAttribution`

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+
+for sample in dataPipe:
+    print(sample)
+    break
+
+# %%
+# Now, let us define few functions to perform tokenization:
+
+def eng_tokenize(text):
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:120]: nit: camel case

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+    """
+    Tokenize an English text and returns list of tokens
+    """
+    return [token.text for token in eng.tokenizer(text)]
+
+def de_tokenize(text):
+    """
+    Tokenize a German text and returns list of tokens
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:128]: nit: "return a list of tokens"

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+# %%
+# Building the vocabulary
+# -----------------------
+# Let us consider an English sentence as the source and a German sentence as the target.
+#
+# Vocabulary can be considered as the set of unique words we have in the dataset.
+# We will build vocabulary for both our source and target now.
+#
+# Let us define a function to get tokens from elements of tuples in the iterator.
+# The comments within the function specifies the need and working of it:
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:148]: This sentence doesn't add too much value. Let's remove.

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+        ## Add <eos> at beginning of each sentence. 2 because the index for <eos> in vocabulary is
+        # 2 as seen in previous section
+        T.AddToken(2, begin=False)
+    )
+    return text_tranform
+
+# %%
+# Now, let us see how to use the above function. The function returns an object of `Transforms`
+# which we will use on our sentence. Let us take a random sentence and check the working of
+# the transform:
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:241]: nit: rephrase to "and check how the transform works."

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    print(indexToString[index], end=" ")
+
+# %%
+# In the above code,:
+#
+#   * At line 2, we take a source sentence from list that we created from dataPipe at line 1
+#   * At line 5, we get a transform based on a source vocabulary and apply it to a tokenized
+#     sentence. Note that transforms take list of words and not a sentence.
+#   * At line 8, we get the mapping of index to string and then use it get the transformed
+#     sentence
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:261]: Remove 2 spaces in front of these bullets so the indentation is consistent with the rest of the tutorial.

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+#
+# Now we will use functions of `dataPipe` to apply transform to all our sentences.
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:263]: rephrase to "DataPipe functions"

File: beginner_source/torchtext_custom_dataset_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+        break
+
+show_some_transformed_sentences(dataPipe)
+# %%
+# In the above output we can observe that the shorter sentences are padded with `<pad>`. Now, we
+# can use this dataPipe while writing our training function.
+#
+# Some parts of this tutorial was inspired from this article:
+# Link: `https://medium.com/@bitdribble/migrate-torchtext-to-the-new-0-9-0-api-1ff1472b5d71\
+# <https://medium.com/@bitdribble/migrate-torchtext-to-the-new-0-9-0-api-1ff1472b5d71>`__.
```
Comment (Nayef211) [inline beginner_source/torchtext_custom_dataset_tutorial.py:383]: nit: directly link the hyperlink to "this article".

**Review by Nayef211 — COMMENTED** [review]
@anp-scp thanks so much for contributing this detailed tutorial. I've left some nits around variable naming and sentence restructuring. Once these are addressed, I'm happy to accept and merge this tutorial.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 1 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 2 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 4 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 5 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested. Also I have added the word "DataPipe" and "DataPipes" in `en-wordlist.txt`.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 6 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 7 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 8 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 9 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 10 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 11 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 12 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

File: beginner_source/torchtext_custom_dataset_tutorial.py (hunk 13 above)
Comment (anp-scp) [reply beginner_source/torchtext_custom_dataset_tutorial.py]: I have made changes as suggested.

**Comment by anp-scp** [comment]
Hi  @Nayef211,

I have updated the tutorial as per the suggestions.

**Review by Nayef211 — APPROVED** [review]
LGTM thanks for updating the tutorial with the suggestions. Will merge once all CI jobs complete! 😄

**Pull request merged**